golem-rib-repl/src/rib_repl.rs

```diff
@@ -144,57 +154,80 @@ impl RibRepl {
     /// This function is exposed for users who want to implement custom REPL loops
     /// or integrate Rib execution into other workflows.
     /// For a built-in REPL loop, see [`Self::run`].
-    pub async fn execute_rib(&mut self, rib: &str) -> Result<Option<RibResult>, RibExecutionError> {
-        if !rib.is_empty() {
-            let rib = rib.strip_suffix(";").unwrap_or(rib);
+    pub async fn execute(
+        &mut self,
+        script_or_command: &str,
+    ) -> Result<Option<RibResult>, RibExecutionError> {
+        let script_or_command = CommandOrExpr::from_str(script_or_command, &self.command_registry)
+            .map_err(RibExecutionError::Custom)?;
 
-            self.update_rib(rib);
+        match script_or_command {
+            CommandOrExpr::Command { args, executor } => {
+                let script = self.repl_state.rib_script();
+                let rib_compiler = self.repl_state.rib_compiler();
 
-            // Add every rib script into the history (in memory) and save it
-            // regardless of whether it compiles or not
-            // History is never used for any progressive compilation or interpretation
-            let _ = self.editor.add_history_entry(rib);
-            let _ = self.editor.save_history(&self.history_file_path);
+                let repl_context =
+                    ReplContext::new(self.printer.as_ref(), &script, &rib_compiler);
 
-            match compile_rib_script(&self.current_rib_program(), &self.repl_state) {
-                Ok(compiler_output) => {
-                    let helper = self.editor.helper_mut().unwrap();
+                executor.run(args.as_str(), &repl_context);
```

Comment:
args here is as much as possible a string. For the most part a REPL command has a single argument however, that may not be case for clients like golem-cli or wasmtime. But they can rely on Clap etc for their arg parsing to form a structured input. This is made possible with the trait `Command` that we expose, which internally talks about a real typed `Input` that's retrieved.

Reply:
When they register their Command that works with varying input and output, it already has an instance of `UntypedCommand` and hence Repl will end up having a collection of `UntypedCommand`